### PR TITLE
Ty 1931 list net training based on dev tool

### PR DESCRIPTION
**Pull request by rustonaut**
Adds support for training on soundgarden user data-frames.

Based on #144 .

Adds following sub-commands to `dev-tool list-net`:

- `convert` prepare samples for training (by converting them from soundgarden user data-frames to a sampels files)
- `train` trains ListNet
- `inspect` inspect ListNet parameter files and/or sample files.

### CODE REVIEW

**Review by rustonaut — COMMENTED** [review]
Self review to check code from a different perspective.

**Comment by rustonaut** [comment]
(I realized I need to update the data on S3, so because of this and the NaN problem some tests will fail.)

File: xayn-ai/src/ltr/mod.rs [hunk 1]
```diff
@@ -116,6 +119,86 @@ impl LtrSystem for ConstLtr {
     }
 }
 
+pub type OwnedSample = (Array2<f32>, Array1<f32>);
+
+/// Creates training data for ListNet from a users history.
+pub fn list_net_training_data_from_history(
+    mut history: &[DocumentHistory],
+) -> Result<Vec<OwnedSample>, Error> {
```
Comment (rustonaut) [inline xayn-ai/src/ltr/mod.rs:127]: This implementation is far from perfect, but will be replaced anyway in a follow-up PR.

**Review by rustonaut — COMMENTED** [review]
Self review + comments for other reviewers.

**Comment by janpetschexain** [comment]
is there a way to split this pr into smaller chunks? 3.5k changed loc is quite a lot oO

**Pull request closed without merging**